prow/ghhook/ghhook.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ghhook
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config/secret"
+
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+)
+
+type Options struct {
+	GitHubOptions    flagutil.GitHubOptions
+	GitHubHookClient github.HookClient
+
+	Repos        flagutil.Strings
+	HookURL      string
+	HMACValue    string
+	HMACPath     string
+	Events       flagutil.Strings
+	ShouldDelete bool
+	Confirm      bool
+}
+
+func GetOptions(fs *flag.FlagSet, args []string) (*Options, error) {
+	o := Options{}
+	o.GitHubOptions.AddFlags(fs)
+	o.Events = flagutil.NewStrings(github.AllHookEvents...)
+	fs.Var(&o.Events, "event", "Receive hooks for the following events, defaults to [\"*\"] (all events)")
+	fs.Var(&o.Repos, "repo", "Add hooks for this org or org/repos")
+	fs.StringVar(&o.HookURL, "hook-url", "", "URL to send hooks")
+	fs.StringVar(&o.HMACPath, "hmac-path", "", "Path to hmac secret")
+	fs.StringVar(&o.HMACValue, "hmac-value", "", "hmac secret value")
+	fs.BoolVar(&o.Confirm, "confirm", false, "Apply changes to github")
+	fs.BoolVar(&o.ShouldDelete, "delete-webhook", false, "Webhook should be deleted")
+	fs.Parse(args)
+
+	if !o.ShouldDelete && o.HMACPath == "" && o.HMACValue == "" {
```

Comment:
Parsing the options from flags and validating their values should be separate functions so that we can still perform option validation in the `hmac` tool even if the options aren't coming from flags.

Reply:
Added a `Validate` function and called it from every place that is needed.